components/supervisor/pkg/supervisor/top_test.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package supervisor
+
+import (
+	"context"
+	"testing"
+)
+
+func TestTopServiceHappyPath(t *testing.T) {
+	ctx := context.Background()
+
+	topService := &TopService{}
+	topService.Observe(ctx)
+
+	<-topService.ready
+
+	if topService == nil {
+		t.Errorf("topService should not be nil")
+	}
+	if topService.data == nil {
+		t.Errorf("topService data should not be nil")
+	}
+	if topService.data.Memory.Used == 0 {
+		t.Errorf("Used Memory should not be zero")
+	}
+	if topService.data.Memory.Limit == 0 {
+		t.Errorf("Total Memory should not be zero")
+	}
+	if topService.data.Cpu.Used == 0 {
+		t.Errorf("Used CPU should not be zero")
+	}
+	if topService.data.Cpu.Limit == 0 {
+		t.Errorf("Total CPU should not be zero")
+	}
+}
+
+// func TestTopServiceFailure(t *testing.T) {
```

Comment:
left over?

Reply:
@akosyakov I am still trying to figure out how to test for failure, either mocking gRPC or injecting the `Top` func dependency.. I would like to add one more test before merging

Reply:
I think you will need to mock `Top` itself via interface: https://gobyexample.com/interfaces I'm not sure whether it is worth a trouble, I tested by removing sock file and gp top was given the last number. In practice I don't expect it to stuck but resolve after sometime to new number.

Reply:
Actually you don't need interface,  `TopService` can have a property `top` which is function, by default you assign `Top` to it. In tests you can provide you own implementation.

Reply:
I've added another test injecting the Top function and this allowed me to simulate failure, the metrics stay the same which demonstrates that works!

I will run some double checks on prev env just to make sure everything works fine there too.